### PR TITLE
Add new config types, improve existing types

### PR DESCRIPTION
**Pull request by djmattyg007**
- Added new Float type. It's very similar to the existing Integer type,
  but accepts floating point numbers, and doesn't allow you to set
  choices.
- Added new Pair type, for when you want to accept a pair of values in a
  single config setting. Pairs default to a pair of strings, but you can
  set either subtype to whatever type you want (including nested pairs).
- Added transformer property to String type. This allows you to apply an
  arbitrary transformation to individual config settings that could
  utimately be anything at all.
- If a transformation is applied to a String config value, it will now
  return a subclass of the built-in str type. This subclass stores the
  original value as well as the transformed value, so that it can be
  re-serialised correctly.
- The "Expected Path" type originally created for the Path type is (for
  now) just a subclass of the "Transformed Value" class, due to the
  duplicate implementation. In future, this should be updated to return
  pathlib Path objects, not strings.
- Made Secret type pass through the transformer property correctly to
  its String parent class.
- Lists can now return a frozenset instead of a tuple, configured by the
  "unique" property.
- Lists now accept an arbitrary subtype, similar to the Pair type. This
  can be any type at all, including Pairs.
- Resolved a couple of todo comments regarding missing tests for the
  List type.
- The Path type will now correctly serialise None to an empty string.

### CODE REVIEW

**Comment by djmattyg007** [comment]
This is based on ideas originally described in #1965 and #1966. Follow-up PRs to actually update the file and http extensions will be created once this PR is merged, to avoid this PR becoming any bigger than it already is.

**Comment by djmattyg007** [comment]
Here are commits for the HTTP and File extensions, demonstrating how the changes in this commit would be used internally.

https://github.com/mopidy/mopidy/commit/f10c93d75910cb380e1a5a52b73a7a5dc61869b3
https://github.com/mopidy/mopidy/commit/f61587ca9fb5f469d284ac51896ec4ea38606cf5

File: mopidy/config/types.py
Comment (jodal) [inline mopidy/config/types.py:252]: I guess this might hit end users, so it should be even more explicit:

```suggestion
            raise ValueError(f"Config value must include {self._separator!r} separator: {raw_value}")
```

File: mopidy/config/types.py
Comment (jodal) [inline mopidy/config/types.py:294]: I think this reads a little bit quicker:

```suggestion
        self._subtype = subtype if subtype else String()
```

File: mopidy/config/types.py
Comment (jodal) [inline mopidy/config/types.py:416]: Do you know of any extensions using it? Given that is prefixed with an underscore, I'd feel quite okay with removing it.

File: tests/config/test_types.py
Comment (jodal) [inline tests/config/test_types.py:267]: `codecs` is in the stdlib, so shouldn't be a problem to move to the top of the file.

**Review by jodal — APPROVED** [review]
Wow, that was quite a lot to review, but it looks solid. :raised_hands: 

This leaves me wondering if there isn't some other format (like TOML) or library (like Pydantic) we could use instead of rolling our own config subsystem. I realize that would be a breaking change and not the most productive thing to spend our spare time on, so I'll settle with a solid test suite :-)

File: mopidy/config/types.py
Comment (djmattyg007) [reply mopidy/config/types.py:294]: I was burned so many times by PHP's crappy ternary operator that I still forget Python has non-crap ternary statements. I agree this is better here :+1:

File: mopidy/config/types.py
Comment (djmattyg007) [reply mopidy/config/types.py]: I've made this change :+1:

File: mopidy/config/types.py
Comment (djmattyg007) [reply mopidy/config/types.py]: I've made this change :+1:

File: tests/config/test_types.py
Comment (djmattyg007) [reply tests/config/test_types.py]: I've made this change :+1:

File: mopidy/config/types.py
Comment (djmattyg007) [reply mopidy/config/types.py:415]: Only the built-in file extension. It was easier to leave it in place rather than delete it, because it gets revamped in this commit I'll be submitting as a follow-up PR:

https://github.com/mopidy/mopidy/commit/07081a7170c4e94c0268972725ffb279b9f2c438

**Comment by djmattyg007** [comment]
> This leaves me wondering if there isn't some other format (like TOML) or library (like Pydantic) we could use instead of rolling our own config subsystem. I realize that would be a breaking change and not the most productive thing to spend our spare time on, so I'll settle with a solid test suite :-)

The thought had definitely occurred to me too, but yeah, it's obviously a significant amount of work and BC breakage.

For the record, I'd be strongly opposed to implementing Pydantic. I'm not a fan of how it requires your domain models to inherit from their base classes, and all the potential problems that can bring.

While I'm not wedded to it, I've had a lot of success with marshmallow for serialisation+deserialisation, and plain old dataclasses from the stdlib for the domain models. In fact, I've even considered whether it would be worth swapping out the existing models implementation for this paradigm. In python 3.10, dataclasses support automatically generating `__slots__`, so you can get the benefit of better memory usage (as well as the long-standing built-in support for immutability) while still avoiding boilerplate.

**Comment by jodal** [comment]
I've done the rounds through attrs, dataclasses, and Pydantic, and have been quite happy with all of them. I'd use dataclasses for any library that doesn't strictly need more. dataclasses with slots and immutability sounds like a good fit for Mopidy's models, but that still leaves the validation part to us.

**Comment by kingosticks** [comment]
I had a random look at some projects in this space yesterday but never used any of them (Mopidy is still my only Python project!). Maybe we should start a new issue/discussion about this topic. Currently I'm not convinced any switch is worth it, we'd always end up having to write extra validators anyway.

**Comment by jodal** [comment]
That discussion is clearly too long-term to block this PR. Do you want to give this a review before we merge, @kingosticks ?

File: mopidy/config/types.py
Comment (kingosticks) [inline mopidy/config/types.py:310]: This may return a different order compared to the value. I think that would be unexpected.

**Review by kingosticks — APPROVED** [review]

**Comment by kingosticks** [comment]
If you rebase you'll pick up the flake8 fix

File: mopidy/config/types.py
Comment (djmattyg007) [reply mopidy/config/types.py:310]: :thinking:

**Comment by djmattyg007** [comment]
I've rebased the branch and added a changelog entry :+1:

File: mopidy/config/types.py
Comment (djmattyg007) [reply mopidy/config/types.py:310]: How much of a concern is it likely to be? The behaviour is opt-in, so I guess we can always adjust the behaviour later based on feedback if absolutely necessary.

Comment (kingosticks) [reply mopidy/config/types.py:310]: If it was documented it'd be better. I just don't like the surprise.

Comment (djmattyg007) [reply mopidy/config/types.py:314]: I've added a comment :+1:

**Pull request merged**
